Fallback to  manifest list when no platform match

cpuguy83 opened the pull request:
relates to / addresses https://github.com/docker/for-linux/issues/1170#issuecomment-750478242


In some cases, in fact many in the wild, an image may have the incorrect
platform on the image config.
This can lead to failures to run an image, particularly when a user
specifies a `--platform`.
Typically what we see in the wild is a manifest list with an an entry
for, as an example, linux/arm64 pointing to an image config that has
linux/amd64 on it.

This change falls back to looking up the manifest list for an image to
see if the manifest list shows the image as the correct one for that
platform.

In order to accomplish this we need to traverse the leases associated
with an image. Each image, if pulled with Docker 20.10, will have the
manifest list stored in the containerd content store with the resource
assigned to a lease keyed on the image ID.
So we look up the lease for the image, then look up the assocated
resources to find the manifest list, then check the manifest list for a
platform match, then ensure that manifest referes to our image config.

This is only used as a fallback when a user specified they want a
particular platform and the image config that we have does not match
that platform.

cpuguy83 committed:
Fallback to  manifest list when no platform match

In some cases, in fact many in the wild, an image may have the incorrect
platform on the image config.
This can lead to failures to run an image, particularly when a user
specifies a `--platform`.
Typically what we see in the wild is a manifest list with an an entry
for, as an example, linux/arm64 pointing to an image config that has
linux/amd64 on it.

This change falls back to looking up the manifest list for an image to
see if the manifest list shows the image as the correct one for that
platform.

In order to accomplish this we need to traverse the leases associated
with an image. Each image, if pulled with Docker 20.10, will have the
manifest list stored in the containerd content store with the resource
assigned to a lease keyed on the image ID.
So we look up the lease for the image, then look up the assocated
resources to find the manifest list, then check the manifest list for a
platform match, then ensure that manifest referes to our image config.

This is only used as a fallback when a user specified they want a
particular platform and the image config that we have does not match
that platform.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

tianon commented on daemon/images/image.go:
```diff
@@ -28,6 +36,113 @@ func (e ErrImageDoesNotExist) Error() string {
+		ra.Close()
+
+		if err != nil {
+			logger.WithError(err).Error("Error reading manifest list for iamge")
```
```suggestion
			logger.WithError(err).Error("Error reading manifest list for image")
```

tianon commented on daemon/images/image.go:
```diff
@@ -44,6 +162,19 @@ func (i *ImageService) GetImage(refOrID string, platform *specs.Platform) (retIm
 		// Note that `platforms.Only` will fuzzy match this for us
 		// For example: an armv6 image will run just fine an an armv7 CPU, without emulation or anything.
 		if !platforms.Only(p).Match(imgPlat) {
+			// Sometimes image variant is not populated due to legacy builders
+			// We still should support falling back here.
+			if imgPlat.OS == platform.OS && imgPlat.Architecture == platform.Architecture && imgPlat.Variant == "" {
+				logrus.WithField("image", refOrID).WithField("platform", platforms.Format(p)).Debug("Image platform cpu variant is not populated, but otherwise it matches what was requested")
+				return
+			}
```
Just to note it somewhere - this is due to a quirk in containerd's platform matching where `platforms.Normalize` is invoked on both the argument to `platforms.Only` _and_ the things it's being compared against, so `linux/arm` gets normalized to `linux/arm/v7` in both cases (and thus fails if you specify `--platform linux/arm/v6` on a single-arch, no-manifest-list image that only has `os` and `architecture`).

Ideally, we'd fix `platforms.Only` in containerd to handle this case better, but it's complicated. :sweat_smile:

tianon reviewed (approved):
Conceptually this seems like a good workaround IMO - the logic in `GetImage` makes sense.

I did my best to review the flow in `manifestMatchesPlatform` and it *seems* sane to me, but I felt like I was stretching my understanding of all the APIs at play in reviewing it (so it should definitely get more eyes :see_no_evil:).

thaJeztah commented:
Discussing in the maintainers meeting; we can remove the check for platform flag, and always warn on mismatch; https://github.com/moby/moby/blob/64bd4485b3a66a597c02c95f5776395e540b2c7c/daemon/create.go#L84-L95

thaJeztah commented:
Also check how it behaves on https://hub.docker.com/r/balenalib/rpi-raspbian, which is not multi-arch, but has wrong arch

cpuguy83 commented:
@thaJeztah That image doesn't have a manifest so this PR doesn't improve anything there as far as I can tell.

cpuguy83 committed:
Move cpu variant checks into platform matcher

Wrap platforms.Only and fallback to our ignore mismatches due to  empty
CPU variants. This just cleans things up and makes the logic re-usable
in other places.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
I updated this to move the CPU variant fallback to a platform matcher that wraps `platforms.Only`.

re: https://github.com/moby/moby/pull/41955#issuecomment-773573286

In the case where a user does specify a platform and there is no match we error out with the same details. I don't think it makes sense to warn and error.

thaJeztah commented:
Looks like there's a linting failiure;

```
daemon/images/image.go:184:3: S1023: redundant `return` statement (gosimple)
		return
 		^
```

thaJeztah commented on daemon/images/image.go:
```diff
@@ -43,18 +163,25 @@ func (i *ImageService) GetImage(refOrID string, platform *specs.Platform) (retIm
+		//   This may be confusing.
+		//   The alternative to this is to return a errdefs.Conflict error with a helpful message, but clients will not be
+		//   able to automatically tell what causes the conflict.
+		retErr = errdefs.NotFound(errors.Errorf("image with reference %s was found but does not match the specified platform: wanted %s, actual: %s", refOrID, platforms.Format(p), platforms.Format(imgPlat)))
+		return
```
```suggestion
		return errdefs.NotFound(errors.Errorf("image with reference %s was found but does not match the specified platform: wanted %s, actual: %s", refOrID, platforms.Format(p), platforms.Format(imgPlat)))
```

thaJeztah replied:
(edited; noticed this was in the `defer`, so no error to return)

cpuguy83 replied:
Fixed

thaJeztah reviewed (approved):
SGTM

tonistiigi reviewed (approved):
SGTM

thaJeztah commented:
Test failiure is unrelated;

```

[2021-02-18T17:11:53.367Z] === FAIL: amd64.integration.image TestRemoveImageGarbageCollector (1.08s)
[2021-02-18T17:11:53.367Z]     remove_unix_test.go:82: assertion failed: error is not nil: Error: No such image: test-garbage-collector
```

that test has the same problem as I fixed in https://github.com/moby/moby/pull/41936
